src/utils/wrappers.py

```diff
@@ -0,0 +1,203 @@
+            logger.info(err)
+            logger.error(f"Exception: {e}", exc_info=True)
+            traceback.print_exc()
+            raise
+
+    return wrapper
+
+
+# def df_change_wrap(func):
+#     """Define a decorator to log the difference between the input and output.
+
+#     This wrapper calculates the differences between the input dataframe
+#     and the output dataframe,
+#     helping to keep a record of changes to the dataframe.
+#     This wrapper should decorate the main
+#     method of a class, e.g. for the `DuplicateWaves` class,
+#     this decorate is applied to the
+#     `duplicate_data()` method. The method this is applied to
+#     must have must have `self` as an
+#     argument and must operate on and then return self.vf_df.
+#     The class as a whole also requires
+#     the table_config attribute to be set.
+#     """
+
+#     @wraps(func)
+#     def call(*args, **kwargs):
+#         """Define the decorator itself."""
+#         self_ = args[0]
+#         pre_df = self_.vf_df
+#         pre_rows, pre_cols = pre_df.shape[0], pre_df.shape[1]
+#         table_config = self_.table_config
+#         post_df = func(*args, **kwargs)
+#         logger.info(f"{func.__name__} changed the df shape as follows:")
+#         df_measure_change(post_df, pre_rows, pre_cols, table_config)
+#         return post_df
+
+#     return call
+
+
+# def df_measure_change(df, rows_before, cols_before, table_config):
+#     """Log the change in a dataframe caused by a function."""
+#     shape = df.shape
+#     rows_after, cols_after = shape[0], shape[1]
+
+#     def _change_direction(before, after):
+#         """Get the direction of the change."""
+#         # Evalutate the direction of change
+#         change = ["gained", "removed"][after < before]
+#         return change
+
+#     row_change = _change_direction(rows_before, rows_after)
+#     col_change = _change_direction(cols_before, cols_after)
+
+#     # Get absolute difference in change
+#     row_diff = abs(rows_after - rows_before)
+#     col_diff = abs(cols_after - cols_before)
+
+#     if table_config == "Table":
+#         # Make table to log the changes in df
+#         tbl = TableLogger(
+#             columns="df Changes,Rows,Columns",
+#             float_format="{:,.2f}".format,
+#             default_colwidth=15,
+#         )
+
+#         tbl("From", rows_before, cols_before)
+#         tbl("To", rows_after, cols_after)
+#         tbl("Change", f"{row_diff} ({row_change})", f"{col_diff} ({col_change})")
+#     elif table_config == "SingleLine":
+#         logger.info(
+#             f"""Difference in rows: {row_diff} ({row_change}),
+#                         Difference in columns {col_diff} ({col_change})"""
+#         )
+#     else:
+#         logger.warning(
+#             """Trouble at mill!!! Mistake in config.
+#                           Either 'Table' or 'SingleLine' must be specified."""
+#         )
```

Comment:
Can you remind me why this is not included and is currently commented out? It's fine if you think this part is beyond the scope of your (already massive) ticket, but please make a new ticket to implement this, plus, extract (cut) the code and put it as detail into the new ticket on Jira.

Reply:
This code was for the table logging which isn't part of ticket 42. I commented it out so the rest of the code would run (I know it isn't best practice). I've since created branch 80 as part of our sprint where I have worked on this code and it runs fine. 

I'll delete this from this branch and you can review the working code in branch-80's pull request.